lib/options: Add mkPackageOption

schuelermine opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Many modules have the option to declare a different package than the default one. These options almost always follow a similar format. This is replicated here.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

schuelermine committed:
lib/options: Add mkPackageOption

This is a squashed commit. These are the original commit messages:

lib/options: Add mkPackageOption

lib/options: Add missing semicolon

lib/options.nix: Make mkPackageOption more complicated

lib/options: Fix indent. & spacing

lib/options.nix: Remove example and align comment

lib/options: ravenous overuse of arguments

lib/options: Format better

lib/options: Add default examplePath decl

lib/options: Make better mkPackageOption function

lib/options: Remove trailing whitespace

lib/options: Improve mkPackageOptions

lib/options: Remove pkgs prefixing
Co-authored-by: pennae <82953136+pennae@users.noreply.github.com>

lib/options: Slim down mkPackageOption further

lib/options: mkPackageOption: Add "pkgs." to example

lib/options: mkPackageOption: Make name & pkgs single arguments

lib/options: mkPackageOption: Swap name & pkgs

lib/options: Remove unnecessary import
Co-authored-by: pennae <82953136+pennae@users.noreply.github.com>

pennae commented on lib/options.nix:
```diff
@@ -99,6 +99,20 @@ rec {
+    mkOption {
+      type = lib.types.package;
+      description = "The ${name} package to use.";
+      inherit default example;
```
this also needs a `defaultText` or the manual builds will fail, and the example should be a `literalExample` to match. at that point we'd have to specify the package not as a `pkgs.X.Y` attrpath, but pass `pkgs` and the attrpath into the function, the latter as a list of strings. at that point we're saving little over writing the option out completely.

(yes, this is not an ideal situation)

schuelermine replied:
Is there a lib function to navigate a path in an attrset?

pennae replied:
there's a few. `getAttrFromPath` is closest, but hard-aborts when it can't find the requested attribute (which is not ideal). `attrByPath` can be given a `(throw "attribute ${...} not found")` default instead of an `abort` default, so that would be better.

schuelermine replied:
Can you add `{ pkgs, lib }: ` to the start of the file to reference nixpkgs itself?

schuelermine replied:
I’ve made a new version, haven’t tested it though, since nixpkgs is so big and I don’t have a local copy (downloading ATM)

schuelermine commented:
O K. 
I did a new implementation… it’s kinda OK, I think?

pennae reviewed (commented):
looks like you're way overcomplicating things :D it's a convenience function, it doesn't have to do everything. a `{ name, pkgs, default /* path in pkgs */, example ? null }` signature would cover at least 90% of uses.

schuelermine commented:
I think this works

schuelermine commented:
OK, I think this is a reasonably convenient function.

pennae commented on lib/options.nix:
```diff
@@ -99,6 +103,33 @@ rec {
+     The argument default is an attribute set path in pkgs.
+  */
+  mkPackageOption =
+    { name ? last default, pkgs, default ? [ name ], example ? null }:
```
one of `name` or `default` should lose their default, otherwise forgetting about both will give you infinite recursion. `name` seems to be a good candidate since many packages referred to by these options have length-1 attrpaths.

pennae commented on lib/options.nix:
```diff
@@ -99,6 +103,33 @@ rec {
+    let
+      default' = if !isList default then
+        [ default ]
+      else if head default == "pkgs" then
+        tail default
```
this part isn't needed, `pkgs.` can be regarded as implicit.

schuelermine replied:
I wanted to make it easier to use so it works if you forget the `pkgs`. I see your point.

pennae commented on lib/options.nix:
```diff
@@ -99,6 +103,33 @@ rec {
+      description = "The ${name} package to use.";
+      default = attrByPath default'
+        (throw "${concatStringsSep "." default'} cannot be found in pkgs") pkgs;
+      defaultText = literalExpression (concatStringsSep "." default');
```
```suggestion
      defaultText = literalExpression ("pkgs." + concatStringsSep "." default');
```

aanderse commented:
Generally I feel indifferent to this function... but I do have one request: Please add documentation stating what overlays are and how they often achieve the same result. I find a large number of PRs which add a `package` option only exist because the author isn't familiar with overlays.

pennae reviewed (commented):
looking good :) are there any modules you think would benefit from this? would be nice to have a few uses in the tree that demonstrate the function, otherwise it'll probably be forgotten again soon.

adding it to the manual would also be really nice.

pennae commented:
@aanderse overlays can be used for many package options, but probably not for all of them. package options are quite useful in container situations where you want one version of the package in a container and another outside of it without the overhead of instantiating nixpkgs multiple times, for example.

aanderse commented:
> @aanderse overlays can be used for many package options, but probably not for all of them. package options are quite useful in container situations where you want one version of the package in a container and another outside of it without the overhead of instantiating nixpkgs multiple times, for example.

I completely agree. And even in the scenario where overlays can be used it can still be desirable to include a `package` option. But let's not lose this opportunity to include some documentation which may teach users something valuable... right?

pennae commented:
agreed! overlay docs would be nice to have as well. :)

schuelermine commented:
@aanderse Overlays can slow down building your entire system because the package is replaced everywhere, not just in user-facing conditions. (i.e. packages that depend on it also use the other version in their dependencies)

schuelermine commented:
Is the documentation autogenerated from the comment or does a doc file need updating as well?

pennae commented:
documentation isn't autogenerated from comments unfortunately, it lives in `nixos/doc/manual` as a set of markdown files. see eg https://nixos.org/manual/nixos/stable/index.html#sec-writing-docs-building-the-manual on how to deal with those :)

pennae commented:
just looking at the signature again and thinking of how it'd be used, as it stands now it doesn't feel like it'd be a great improvement over the current verbosity. you'd still have quite a block:
```
package = mkPackageOption {
  name = "Package";
  inherit pkgs;
  default = "package";
};
```
this would generate the `defaultText` you wouldn't get for free otherwise of course, but it's still a bit too much to call it "convenient". maybe if it were
```
package = mkPackageOption pkgs "Package" {
  default = "package";
};
```

don't know, honestly. maybe attacking `user` and `group` options first would be better to establish a pattern of how these things could look and what they'd do. :thinking: any thoughts?

schuelermine commented:
@pennae I implemented that

pennae commented on lib/options.nix:
```diff
@@ -35,6 +36,9 @@ let
   inherit (lib.types)
     mkOptionType
     ;
+  inherit (lib.lists)
+    last
+    ;
```
```suggestion
```
no longer used.

pennae reviewed (commented):
the function looks pretty convenient now :)

have you found your way around the docs? this really should be in there.

schuelermine commented:
1) No, I haven’t found precisely where I should put this in the docs
2) Does nixpkgs recommend a nix formatter? I have nixfmt setup to auto-format on save usually and it messes up all the files…

schuelermine committed:
nixos/docs/option-declarations: Document mkEnableOption and mkPackageOption

This is a squashed commit. These are the original commit messages:

lib/option: Improve comment

better comment

Update documentation

Updated nixos/doc/manual/development/options-declarations.md with info on mkEnableOption and mkPackageOption.
Updated the comment on mkEnableOption in lib/options.nix

remove trailing whitespace

nixos/doc/option-declarations: Update IDs & formatting

nixos/docs/option-declarations: Escape angle brackets

Build DB from MD

(Amended) Fix typo
Co-authored-by: pennae <82953136+pennae@users.noreply.github.com>

(Amended) Build DB from MD (again)

schuelermine commented:
I was searching through ^/doc/functions/library, and there’s attrsets.xml, but nothing else… and ^/doc/functions/library.xml says “<xi:include href="./library/**generated**/…>”, which suggests things are autogenerated.
And I searched through the code for some phrases from the manual about some of the functions and I can’t actually find those exact phrases anywhere.

schuelermine commented:
If you’re talking about the NixOS manual, `mkEnableOption` isn’t even documented there…

pennae commented:
`nixos/doc/manual/development/option-declarations.section.md` is probably a good place to put docs for this. if you're up to `mkEnableOption` would fit in there too :) the xml files are generated from md with `md-to-db.sh`, and from there you can build the manual as usual.

nixpkgs does not currently recommend any formatter (or even format). as long as new code roughly matches what surrounds it things are pretty much fine.

schuelermine commented:
Also, it seems to me that the docs for the nixpkgs library itself in ^/doc/functions/… are auto-generated (at least the part containing the `lib.options` functions), and I’ve tried to rebuild them by doing `nix-shell` and then `make all`, but the built docs (which include `mkEnableOption`) never change. Do I need to format my comment specially? (I’m pretty sure the comment is the source of the text, no-where else is it to be found)

pennae commented:
some documentation is generated, but module documentation isn't (see `doc/doc-support/lib-function-docs.nix` for details). modules, options, types are written rather than generated at this point.

schuelermine commented:
OK, but how do I get the autogenerated library docs to update? `make all` inside `nix-shell` in `doc` doesn’t work.

schuelermine commented:
(sorry for force push, I forgot to add the comment)

schuelermine commented:
OK so this manual thing evidently isn’t working out. Can anybody help with the formatting/syntax?

pennae commented:
the docs texts looks good! :+1: 

you can regenerate the xml files with `nixos/doc/manual/md-to-db.sh`. after that you can build the manual as you had tried, or build it as part of a full system.

schuelermine commented:
OK, I’m now getting a weird error when building the manual.

<details>

<summary> (Click to expand) Abridged output of <tt>make</tt> </summary>

```none
…
./man-nixos-version.xml: OK
The component XML files aren't valid as subcomponents.

Trying the whole document just in case.

----

Doing a combined check...
from_md/development/option-declarations.section.xml:114: parser error : Opening and ending tag mismatch: name line 114 and quote
        <name>.</quote>.
                       ^
from_md/development/option-declarations.section.xml:115: parser error : Opening and ending tag mismatch: quote line 113 and para
      </para>
             ^
from_md/development/option-declarations.section.xml:157: parser error : Opening and ending tag mismatch: name line 157 and quote
          description <quote>The <name> package to use.</quote>. You can
                                                               ^
from_md/development/option-declarations.section.xml:160: parser error : Opening and ending tag mismatch: quote line 157 and para
        </para>
               ^
from_md/development/option-declarations.section.xml:321: parser error : Opening and ending tag mismatch: para line 155 and section
      </section>
                ^
from_md/development/option-declarations.section.xml:323: parser error : Opening and ending tag mismatch: para line 112 and section
  </section>
            ^
from_md/development/option-declarations.section.xml:325: parser error : Premature end of data in tag section line 100

^
from_md/development/writing-modules.chapter.xml:187: element include: XInclude error : could not load from_md/development/option-declarations.section.xml, and no fallback was found
./manual-combined.xml: Not valid:
Line 12168:
 12165	 };
 12166	}
 12167	</programlisting>
 12168	 <xi:include href="option-declarations.section.xml"/>
 12169	 <section xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="sec-option-types">
 12170	 <title>Options Types</title>
 12171	 <para>
     error: element "xi:include" not 
    allowed here; expected the element 
    end-tag or element "address", "anchor", 
    "annotation", "bibliography", 
    "bibliolist", "blockquote", 
    "bridgehead", "calloutlist", "caution", 
    "classsynopsis", "cmdsynopsis", 
    "constraintdef", "constructorsynopsis", 
    "destructorsynopsis", "epigraph", 
    "equation", "example", "fieldsynopsis", 
    "figure", "formalpara", "funcsynopsis", 
    "glossary", "glosslist", "important", 
    "index", "indexterm", 
    "informalequation", "informalexample", 
    "informalfigure", "informaltable", 
    "itemizedlist", "literallayout", 
    "mediaobject", "methodsynopsis", 
    "msgset", "note", "orderedlist", 
    "para", "procedure", "productionset", 
    "programlisting", "programlistingco", 
    "qandaset", "refentry", "remark", 
    "revhistory", "screen", "screenco", 
    "screenshot", "sect1", "section", 
    "segmentedlist", "sidebar", "simpara", 
    "simplelist", "simplesect", "synopsis", 
    "table", "task", "tip", "toc", 
    "variablelist" or "warning"
```

</details>

schuelermine commented:
Ok, I think that's done. Should the commits be squashed?

pennae commented on lib/options.nix:
```diff
@@ -99,6 +100,49 @@ rec {
+       => { _type = "option"; default = «derivation /nix/store/jxx55cxsjrf8kyh3fp2ya17q99w7541r-ghc-8.10.7.drv»; defaultText = { ... }; description = "The GHC package to use."; example = { ... }; type = { ... }; }
+  */
+  mkPackageOption =
+    # Package set (a specific version of nixpkgs)
+    pkgs:
+      # Name for the package, shown in option description
+      name:
+      { default ? [ name ], example ? null }:
```
just a tiny stilistic nit if your going to go over it anyway:
```suggestion
    pkgs: # Package set (a specific version of nixpkgs)
    name: # Name for the package, shown in option description
    { default ? [ name ], example ? null }:
```

pennae commented on nixos/doc/manual/development/option-declarations.section.md:
````diff
@@ -57,6 +57,80 @@ The function `mkOption` accepts the following arguments.
+mkPackageOption pkgs "name" { default = [ "path" "in" "pkgs" ]; example = "literal example"; }
+```
+
+Creaties an Option attribute set for an option that specifies the package a module should use for some purpose.
````
```suggestion
Creates an Option attribute set for an option that specifies the package a module should use for some purpose.
```

schuelermine replied:
I re-applied this

pennae reviewed (commented):
yes, squashing would be preferrable since there's really only two changes here (docs for mkEnableOption, and the rest).

schuelermine commented:
Is there a way to squash non-adjacent/-consecutive commits into multiple separate commits? Or should this just be one commit?﻿

schuelermine commented:
OK, I’ve squashed it down to two commits

schuelermine commented:
Somewhere along the way the typo fix got lost. It’s back.

schuelermine commented:
( I probably just forgot to pull )

schuelermine commented:
Actually, I got an idea for a new feature. Hold on.

schuelermine commented:
I think this could be improved further, I have it on patch-mkPackageOption-extra, but maybe that’s beyond the scope of this pull request or this function.

pennae commented:
improvements can always be made if and when they turn out to be helpful :) looks good as is. let's hope users will appear, then we'll have a base of experience for improvements.
